### PR TITLE
feat(sdk): support optional artifact inputs (and improve support for optional parameter inputs)

### PR DESCRIPTION
**Pull request by connor-mccarthy**
**Description of your changes:**
Adds support for optional artifact inputs to maintain feature parity with v1.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the 
pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by connor-mccarthy** [comment]
/hold
Rebase and merge after #8574 and #8575 merge
Wait until new kfp-pipeline-spec release

**Comment by connor-mccarthy** [comment]
/hold

Wait to merge until next release of kfp-pipeline-spec library, which includes `is_optional` field for artifact inputs

File: sdk/python/kfp/compiler/pipeline_spec_builder.py [hunk 1]
```diff
@@ -330,56 +330,18 @@ def build_component_spec_for_task(
...
+                f'PipelineTaskFinalStatus can only be used in an exit task. Parameter {input_name} of a non exit task has type PipelineTaskFinalStatus.'
+            )
+    component_spec = _build_component_spec_from_component_spec_structure(
+        task.component_spec)
+    component_spec.executor_label = utils.sanitize_executor_label(task.name)
     return component_spec
 
 
-# TODO(chensun): merge with build_component_spec_for_task
 def _build_component_spec_from_component_spec_structure(
```
Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/pipeline_spec_builder.py:345]: https://docs.google.com/spreadsheets/d/1HbBYceT4hxMtr3GXSyE8hgdN3ZjWAEuVdcUxUT6Ju-U/edit#gid=1014673294

Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/pipeline_spec_builder.py:345]: Primary changes are here

File: sdk/python/test_data/components/container_with_if_placeholder.yaml [hunk 2]
```diff
@@ -9,6 +9,7 @@ components:
       parameters:
         optional_input:
           defaultValue: default
+          isOptional: true
```
Comment (chensun) [inline sdk/python/test_data/components/container_with_if_placeholder.yaml:12]: It seems like all the test golden files updates are for optional parameter inputs, which isn't really improving support but fixing a bug: information lost for optional input parameter with None default value when preserving into YAML.
It's fine to go ahead with this change. But ideally, in the future we should keep the changes separate: fixing bug for optional input parameters, and add support for optional input artifacts.

**Review by chensun — COMMENTED** [review]
/lgtm

Feel free to merge when it's ready to go. Thanks!

**Pull request closed without merging**